### PR TITLE
first test and local_backend with default storage

### PR DESCRIPTION
**Pull request by fhahn**
Hi,

I've added a first test for the LocalUploadBackend and modified LocalUploadBackend to use default_storage.

Because of a bug in Django's test client, you'll need a recent version of Django's trunk to run the test.

LocalUploadBackend uses the Django File [0] and Storages API [1] to save uploaded files, so it should be possible to switch the default storage (django-storages [2] provides a lot of different storage backends) and  upload files to that storage with the LocalUploadBackend. I've only tested it with the FileSystemStorage (the "default storage") so far, but I'm going to test it with other backends soon.

What do you think of the changes?
Should I move the default storage stuff to a new backend and leave the old LocalUploadBackend as it is?

Cheers,
Flo

[0] https://docs.djangoproject.com/en/1.3/ref/files/file/
[1] https://docs.djangoproject.com/en/1.3/ref/files/storage/
[2] https://bitbucket.org/david/django-storages/src

### CODE REVIEW

**Comment by skoczen** [comment]
Hey Flo,

Love the idea of this, will look through it in detail to get you a proper response shortly. 

Generally - I like the idea of using default_storage, but I'd want to keep backwards compatibility. I believe ajaxuploader works back to at least 1.2 right now, and it'd be nice to keep it that way.

How about implementing a conditional import, that requires django-storages if it can't be imported from django?

Love pull requests with tests!

-Steven

**Comment by fhahn** [comment]
Hey Steven,

I think we should provide an easy way to change the UploadBackend used by the AjaxFileUploader view (e.g. with setting the backend ins settings.py) and move the Django storage based backend from LocalUploadBackend to something StorageUploadBackend. 

So everything will just keep working at the moment, but people can start using the StorageUploadBackend. 

What do you think?

**Comment by skoczen** [comment]
Hey Flo,

+1 to moving the storage-based one to StorageUploadBackend.

+0 to the settings.py, on "explicit is better than implicit" grounds.  Since all the other settings for a subclass are in one place, I'd rather specify the backend there as well, and not rely on magical lookups.   As a part of a bundled url/view package, I could see using the settings - but don't see how they make things clearer in this case.  If you've got an argument for it, I'd love to hear it.

**Comment by fhahn** [comment]
Hi Steven,

sorry for the long delay, but things are quite busy around here.

I'll move the backend as soon as I can, so this you'll hopefully be able to merge this branch (and the tests) soon.

now to the settings thing:
I completely overlooked the backend parameter in AjaxFileUploader.**init**, so just forget the things I said about the settings.py stuff. That's a neat solution to solve the problem!

Cheers,
Flo

**Comment by skoczen** [comment]
Hey Flo,

Not a worry - things are crazy over here as well.  No offense intended if it takes me a bit to merge your changes when they do come over.

Glad the init works for you, look forward to the pull!

-S

**Comment by fhahn** [comment]
I finally had time to move the code around. The backend now lives in backends/default_storage.py and the tests in tests/default_storage_backend.py.

Cheers, 
Flo

File: ajaxuploader/backends/local.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def setup(self, filename):
             os.makedirs(os.path.realpath(os.path.dirname(self._path)))
         except:
             pass
-        self._dest = BufferedWriter(FileIO(self._path, "w"))
```
Comment (skoczen) [inline ajaxuploader/backends/local.py:18]: This looks wrong to me.  Tab indent error, or am I missing something?

**Comment by skoczen** [comment]
Other than the above question, this looks great to me.

Can you toss in documentation for your backend into the readme?

Once that's there, I'm merging it in! 

Thanks,
-Steven

**Comment by fhahn** [comment]
I've fixed the indent error and added some information about the backend to the readme

Cheers and merry christmas,
Flo

**Comment by skoczen** [comment]
Looks fantastic. Thanks for keeping on this - it definitely makes ajax-uploader better.

Cheers & happy holidays to you as well,

-S

**Pull request merged**